File: library-core/src/main/java/com/mikepenz/fastadapter/utils/DefaultItemListImpl.java
package com.mikepenz.fastadapter.utils;

import com.mikepenz.fastadapter.IAdapterNotifier;
import com.mikepenz.fastadapter.IItem;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nullable;

/**
 * The default item list implementation
 */

public class DefaultItemListImpl<Item extends IItem> extends DefaultItemList<Item> {

    protected List<Item> mItems;

    public DefaultItemListImpl() {
        this(new ArrayList<Item>());
    }

    public DefaultItemListImpl(List<Item> items) {
        this.mItems = items;
    }

    @Override
    public Item get(int position) {
        return mItems.get(position);
    }

    @Override
    public List<Item> getItems() {
        return mItems;
    }

    @Override
    public int getAdapterPosition(long identifier) {
        for (int i = 0, size = mItems.size(); i < size; i++) {
            if (mItems.get(i).getIdentifier() == identifier) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public void remove(int position, int preItemCount) {
        mItems.remove(position - preItemCount);
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemRemoved(position);
        }
    }

    @Override
    public void removeRange(int position, int itemCount, int preItemCount) {
        //global position to relative
        int length = mItems.size();
        //make sure we do not delete to many items
        int saveItemCount = Math.min(itemCount, length - position + preItemCount);

        for (int i = 0; i < saveItemCount; i++) {
            mItems.remove(position - preItemCount);
        }
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemRangeRemoved(position, saveItemCount);
        }
    }

    @Override
    public void move(int fromPosition, int toPosition, int preItemCount) {
        Item item = mItems.get(fromPosition - preItemCount);
        mItems.remove(fromPosition - preItemCount);
        mItems.add(toPosition - preItemCount, item);
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemMoved(fromPosition, toPosition);
        }
    }

    @Override
    public int size() {
        return mItems.size();
    }

    @Override
    public void clear(int preItemCount) {
        int size = mItems.size();
        mItems.clear();
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemRangeRemoved(preItemCount, size);
        }
    }

    @Override
    public boolean isEmpty() {
        return mItems.isEmpty();
    }

    @Override
    public void set(int position, Item item, int preItemCount) {
        mItems.set(position - preItemCount, item);
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemChanged(position);
        }
    }

    @Override
    public void addAll(List<Item> items, int preItemCount) {
        int countBefore = mItems.size();
        mItems.addAll(items);
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemRangeInserted(preItemCount + countBefore, items.size());
        }
    }

    @Override
    public void addAll(int position, List<Item> items, int preItemCount) {
        mItems.addAll(position - preItemCount, items);
        if (getFastAdapter() != null) {
            getFastAdapter().notifyAdapterItemRangeInserted(position, items.size());
        }
    }

    @Override
    public void set(List<Item> items, int preItemCount, @Nullable IAdapterNotifier adapterNotifier) {
        //get sizes
        int newItemsCount = items.size();
        int previousItemsCount = mItems.size();

        //make sure the new items list is not a reference of the already mItems list
        if (items != mItems) {
            //remove all previous items
            if (!mItems.isEmpty()) {
                mItems.clear();
            }

            //add all new items to the list
            mItems.addAll(items);
        }
        if (getFastAdapter() == null) return;
        //now properly notify the adapter about the changes
        if (adapterNotifier == null) {
            adapterNotifier = IAdapterNotifier.DEFAULT;
        }
        adapterNotifier.notify(getFastAdapter(), newItemsCount, previousItemsCount, preItemCount);
    }

    @Override
    public void setNewList(List<Item> items, boolean notify) {
        mItems = new ArrayList<>(items);
        if (getFastAdapter() != null && notify) {
            getFastAdapter().notifyAdapterDataSetChanged();
        }
    }
}

File: library-core/src/main/java/com/mikepenz/fastadapter/utils/DefaultItemListImpl.kt
package com.mikepenz.fastadapter.utils

import com.mikepenz.fastadapter.GenericItem
import com.mikepenz.fastadapter.IAdapterNotifier
import java.util.*
import kotlin.math.min

/**
 * The default item list implementation
 */
open class DefaultItemListImpl<Item : GenericItem> @JvmOverloads constructor(
        @Suppress("ConstructorParameterNaming")
        protected var _items: MutableList<Item> = ArrayList()
) : DefaultItemList<Item>() {

    override val items: MutableList<Item>
        get() = _items

    override val isEmpty: Boolean
        get() = _items.isEmpty()

    override fun get(position: Int): Item {
        return _items[position]
    }

    override fun getAdapterPosition(identifier: Long): Int =
            _items.indexOfFirst { it.identifier == identifier }

    override fun remove(position: Int, preItemCount: Int) {
        _items.removeAt(position - preItemCount)
        fastAdapter?.notifyAdapterItemRemoved(position)
    }

    override fun removeRange(position: Int, itemCount: Int, preItemCount: Int) {
        //global position to relative
        val length = _items.size
        //make sure we do not delete too many items
        val saveItemCount = min(itemCount, length - position + preItemCount)
        for (i in 0 until saveItemCount) {
            _items.removeAt(position - preItemCount)
        }
        fastAdapter?.notifyAdapterItemRangeRemoved(position, saveItemCount)
    }

    override fun move(fromPosition: Int, toPosition: Int, preItemCount: Int) {
        val item = _items[fromPosition - preItemCount]
        _items.removeAt(fromPosition - preItemCount)
        _items.add(toPosition - preItemCount, item)
        fastAdapter?.notifyAdapterItemMoved(fromPosition, toPosition)
    }

    override fun size(): Int {
        return _items.size
    }

    override fun clear(preItemCount: Int) {
        val size = _items.size
        _items.clear()
        fastAdapter?.notifyAdapterItemRangeRemoved(preItemCount, size)
    }

    override fun set(position: Int, item: Item, preItemCount: Int) {
        _items[position - preItemCount] = item
        fastAdapter?.notifyAdapterItemChanged(position)
    }

    override fun addAll(items: List<Item>, preItemCount: Int) {
        val countBefore = _items.size
        _items.addAll(items)
        fastAdapter?.notifyAdapterItemRangeInserted(preItemCount + countBefore, items.size)
    }

    override fun addAll(position: Int, items: List<Item>, preItemCount: Int) {
        _items.addAll(position - preItemCount, items)
        fastAdapter?.notifyAdapterItemRangeInserted(position, items.size)
    }

    override fun set(items: List<Item>, preItemCount: Int, adapterNotifier: IAdapterNotifier?) {
        //get sizes
        val newItemsCount = items.size
        val previousItemsCount = _items.size

        //make sure the new items list is not a reference of the already mItems list
        if (items !== _items) {
            //remove all previous items
            if (_items.isNotEmpty()) {
                _items.clear()
            }

            //add all new items to the list
            _items.addAll(items)
        }
        fastAdapter?.let { fastAdapter ->
            //now properly notify the adapter about the changes
            (adapterNotifier ?: IAdapterNotifier.DEFAULT).notify(
                    fastAdapter,
                    newItemsCount,
                    previousItemsCount,
                    preItemCount
            )
        }
    }

    override fun setNewList(items: List<Item>, notify: Boolean) {
        _items = ArrayList(items)
        if (notify) {
            fastAdapter?.notifyAdapterDataSetChanged()
        }
    }
}
